ValkeyLDAP.md

```diff
@@ -0,0 +1,271 @@
+- TLS 1.2+ required for AD communication
+- Certificate validation
+- Secure credential storage
+- Encrypted cache entries
+
+### Maintainability
+
+- Clear separation of concerns
+- Modular design for easy updates and inclusion of new protocols
+- Comprehensive logging and monitoring
+- Simple configuration management
+- Easy troubleshooting and debugging
+
+### Compatibility
+
+- Support for various Valkey deployment models
+- Backward compatibility with existing Valkey features
+- Integration with different LDAP providers
+- Support for multiple authentication methods
+- Compatibility with existing Valkey clients
+
+### High Availabilty
+
+- Multiple identity provider server support
+- Connection failover
+- Cache replication
+- Graceful degradation
+
+### Error Handling
+
+- Invalid credentials
+- Network timeouts
+- LDAP server unavailable
+- Cache corruption
+- Configuration errors
+- Group resolution failures
+
+## Specification
+
+### Authentication flow
+
+1. Client sends AUTH command
+2. Module extracts username/password
+3. Check credential cache
+4. If not cached:
+    - connect to LDAP server
+    - perform bind operation
+    - search for user group membership
+    - cache successful result
+    - set TTL
+5. Pass AUTH/NOAUTH to Valkey core
+6. Return auth response
+
+### On module load
+
+- Load configuration from Valkey config or default values
+- Initialize data structures and connection pools
+- Register command hooks with valkey core
+- Establish initial LDAP server connections
+- Load cached credentials if persistence enabled
+- Register monitoring callbacks
+
+### Commands 
+
+1 LDAP.CONFIG SET <parameter> <value>
+
+- Sets configuration parameters for the LDAP authentication module
+- Parameters include server addresses, timeouts, cache settings, etc.
+- Returns simple string reply: OK on success
+
+2 LDAP.CONFIG GET <parameter>
+
+- Gets current configuration for specified parameter
+- Returns bulk string reply: parameter value
+
+3 LDAP.RELOAD
+
+- Reloads configuration from config file
+- Reestablishes connections with LDAP servers
+- Returns simple string reply: OK on success
+
+4 LDAP.STATUS
+
+- Returns health and statistics information
+- Format: nested hash with connection states, cache stats, auth counts
+
+5 LDAP.FLUSHCACHE
+
+- Clears the credential cache
+- Options to flush all or specific users
+- Returns integer reply: number of entries flushed
+
+### Valkey Command Hooks
+
+#### Authentication Command Hook
+
+- Intercepts AUTH command
+- Processes credentials through LDAP binding
+- Maintains original Valkey auth functionality for local accounts
+
+#### ACL Command Hook
+
+- Integrates with ACL commands
+- Maps LDAP groups to Valkey ACLs
+- Preserves ACL inheritance and evaluation logic
+
+### Internal data structures
+
+#### Configuration Store
+
+- Hash table for configuration parameters
+- Persistent across restarts via RDB or config file
+- Thread-safe access with read-write locks
```

Comment:
Why thread-safe? Why do we need threads at all?

Can't we use non-blocking IO and use Valkey's event loop? There's the event loop API to do that: `ValkeyModule_EventLoopAdd`, etc.

Reply:
There shouldn't be any cross thread memory accesses. Inside AWS we almost always run a second thread with a separate event loop for not user data (like doing authentication requests) just so we aren't stealing resources for command processing. This is similar to how we offload all of the IO.

Reply:
I guess there are pros and cons. If you run a small node on a single-threaded virtual environment, it's bad to use threads. Context switches slow down a lot. Scaling horizontally with many single-threaded nodes is how you can get most out of the CPUs.